### ISSUE TITLE
Order CSV Export Issue with MS Excel

### ISSUE THREAD

**Issue by rain2o** [magento/magento2#12473]
### Preconditions

1. Magento 2.1.10
2. PHP 7.0

### Steps to reproduce

1. In Admin navigate to the Sales Order grid
2. Filter the orders according to your needs
3. Export as CSV
4. Attempt to open the file in Microsoft Excel

### Expected result

1. The file should open as a spreadsheet.

### Actual result

1. Excel prompts with an error - "Excel has detected that 'export.csv' is a SYLK file, but cannot load it. Either the file has errors or it is not a SYLK file format. Click OK to try to open the file in a different format."
2. When clicking OK it works for some users and not others. So far it seems to work for Excel for Mac but does **not** work for Excel on Windows. 

Basically Excel sees the first two characters as "ID" and assumed it is a SYLK format. The easiest way to fix this is to change the ID column to have a different label. Locally I changed it to "Order ID" and that fixed the issue. It is a small change for a unique use-case, but I hope we can implement this at some point. 

Technically this is a bug with Microsoft Excel, but as it is highly unlikely for them to implement a fix for this as it has been around for a long time I believe, I am submitting this bug as an attempt to work around the unique requirements of Excel.

**Comment by AirmanAJK** [comment]
It is also worth noting that if a CSV file opened in Excel does not start with a standard UTF8 byte order mark (\uFEFF), excel will not consider the file encoding as Unicode. This means any special symbols you might have in your product description fields will get scrambled.

Here is an easy test to reproduce:
Edit a product description field by appending the following: ❤ ☀ ☆ ☂ ☻ ♞ ☯ ☭ ☢ 
Export the product to a CSV file.
Open in excel and notice the resulting value: â¤ â˜€ â˜† â˜‚ â˜» â™ž â˜¯ â˜­ â˜¢

If you were to update the CSV file in Notepad++ by changing the encoding to "UTF-8-BOM", save, and reopen in excel, the proper characters will now be displayed.

**Comment by orlangur** [comment]
It is a known Excel issue for both Windows and Mac: https://support.microsoft.com/en-us/help/323626/-sylk-file-format-is-not-valid-error-message-when-you-open-file But seems like only versions older than 2004 are affected. Which version of Excel did you observe it on?

As a workaround I believe adding apostrophe in the common csv generator is the best choice, no need to rename particular columns.

**Comment by AirmanAJK** [comment]
Can confirm the ID as the first column issue is occurring in Excel 2016

**Comment by orlangur** [comment]
Oh :) Then this help article is misleading. Probably not going to fix it due to backward compatibility or something.

Could you please check also that workaround with apostrophe described there works?

**Comment by rain2o** [comment]
I am using Excel 2016 for Mac and seeing the error. 

My workaround was to rename the column to "Order ID". Adding the apostrophe should work as well. I'll have to test that. 

I have been running through xDebug and realized that the column starts out with ID in quotes as a column header (as it should be), but the downloaded file no longer has quotes around ID. I opened it directly in a text editor and it is the only column heading without quotes. Putting quotes around it fixes the issue as well. I'm not positive what is causing that to happen yet but I'm still digging in.

**Comment by rain2o** [comment]
Ah, right. ID isn't in quotes because `fputcsv` only encloses columns in quotes if they are more than one word. So that explains that. 

@orlangur In response to your recommendation to add the apostrophe, that seems to also work. However personally I would prefer to rename the column "ID" to something less generic like Order ID or Increment ID as opposed to adding an apostrophe to the beginning of every single CSV that M2 generates. I could picture our clients reporting that as a bug if every CSV they export has an apostrophe at the beginning.

**Comment by orlangur** [comment]
> would prefer to rename the column "ID" to something less generic like Order ID or Increment ID as opposed to adding an apostrophe to the beginning of every single CSV that M2 generates

That would require code of every affected grid to be spoiled with Excel-specific hack. And also new grids would be affected without such hack.

> I could picture our clients reporting that as a bug if every CSV they export has an apostrophe at the beginning.

Only every CSV starting with "ID" (case-sensitive). That is official Microsoft recommendation after all, your clients cannot call a bug something that does not break anything.

**Comment by rain2o** [comment]
Valid points @orlangur . Maybe I misunderstood your original workaround solution then. The way I understood it the apostrophe would be added to all CSV's that are generated in M2. So you are recommending adding logic to the header generation portion that checks the first column for the value "ID" and adds the apostrophe then? 

This is definitely a very specific use-case, and technically working around a bug in Excel. But as this has been a known issue since 2004 at least, and still is in 2016 it's become pretty apparent that Microsoft has no intention of fixing it.

**Comment by orlangur** [comment]
> adding logic to the header generation portion that checks the first column for the value "ID" and adds the apostrophe then?

Yeah, exactly. Link to Microsoft help in place of such workaround wouldn't hurt. Every CSV not starting with ID will remain unchanged.

**Comment by magento-engcom-team** [comment]
Hi @rain2o. Thank you for the report. We are moving your feature request to the special project. You can track this issue here: https://github.com/magento-engcom/community-features/issues/11
